pypuf/learner/lr.py

```diff
@@ -0,0 +1,212 @@
+            model_gradient = model_gradient_xor
+        elif compare_functions(self.combiner == LTFArray.combiner_ip_mod2):
+            model_gradient = model_gradient_ip_mod2
+        else:
+            raise Exception('No gradient function known for combiner %s' % self.combiner)
+
+        ret = array([
+            dot(
+                self.derivative
+                *
+                # estimated responses of the l-th LTF, derived from the real total value
+                # and the (k-1) modelled values
+                combined_model_responses / model_responses[:,l],
+                self.transformed_challenges[:,l]  # all challenges to the l-th Arbiter chain
+            )
+            for l in range(self.k)
+        ])
+        return ret
+
+    def learn(self):
+        # let numpy raise exceptions
+        seterr(all='raise')
+
+        # we start with a random model
+        model = LTFArray(
+            weight_array=LTFArray.normal_weights(self.n, self.k, self.mu, self.sigma),
+            transform=self.transformation,
+            combiner=self.combiner,
+        )
+
+        updater = self.RPropModelUpdate(model)
+
+        converged = False
+        distance = 1
+        self.iteration_count = 0
+        while not converged and distance > .01 and self.iteration_count < self.iteration_limit:
+            stderr.write('\riter %5i, dist % 2.4f         ' % (self.iteration_count, distance))
+            self.iteration_count += 1
+
+            # compute gradient & update model
+            gradient = self.gradient(model)
+            self.print_status(model)
+            model.weight_array += updater.update(gradient)
+
+            # check convergence
+            converged = (
+                [0] * self.n ==
+                around(gradient, decimals=self.convergence_decimals)
+            ).all()
+
+            # check accuracy
+            distance = (self.training_set.N - count_nonzero(self.training_set.responses == self.sign_combined_model_responses)) / self.training_set.N
+            self.min_distance = min(distance, self.min_distance)
+
+        if not converged and distance > .01:
+            print('WARNING, MODEL DID NOT CONVERGE')
+
+        return model
+
+    def print_status(self, model):
+        return # debug output disabled
+        print('\nLR Training Set vs. Model Responses vs. Class Probability\n----------------------------------------------------------')
+        for (idx,c) in enumerate(self.set.challenges):
+            print(' '.join(['%+1d' % ci for ci in c]) +
+                  ' => ' +
+                  ('+1' if self.set.responses[idx] == 1 else '-1') +
+                  '    ' +
+                  '  '.join([('%+ 7.2f' % mr) for mr in model.ltf_eval([c])[0]]) +
+                  '    ' +
+                  ('% 10.5f' % self.class_probability[idx])
+                  )
+
+    def __getstate__(self):
+        print(self.__dict__)
```

Comment:
GitHub doesn't seem happy with no newline at the end of the file. Is that somehow important?